### DISCUSSION TITLE
Ability to extend the default fence codeblock to include meta options.

### DISCUSSION THREAD

**Discussion by Lofty-Brambles** [markdoc/markdoc#318] [Ideas]
So, from the default parsing method extended from markdoc, it reads the "language" tag from the start of the fence to first space. https://github.com/markdoc/markdoc/blob/dcc4d4633ece101c46cefbe550e3120183cc3a45/src/parser.ts#L52
I understand that markdoc aim's to be precisely compliant with markdown, but are there any issues with adding that entire line's information, to a "meta" attribute, which does not need to be rendered by default? That ensures an easy to parse area for attributes to the fences while remaining compatible with markdown, and would allow the fences to be used as a more flexible code-block component, which's what it's meant to do, over adding yet another code tag for any simple function, like highlighting a particular line.

**Comment by rpaul-stripe** [comment]
Markdoc has built-in support for adding custom attributes to fences. If you add an annotation to that opening line, it will parse the attributes and expose them on the fence's AST node:

~~~
```javascript {% foo="bar" %}
console.log("This is code in a fenced code block");
```
~~~

You can use this functionality to customize the behavior of fence nodes and add metadata that is accessible in transform functions. You can also override the schema for a fence node to declare the attributes so that the attribute values are validated. We use this functionality at Stripe to support line highlighting and other specialized behaviors we want in fences.

Does that meet your needs? We probably could expose the raw unparsed value of the line if there's really a compelling use case, but I want to encourage people to take advantage of the attribute system instead if possible.

**Reply by 66Leo66** [reply]
@rpaul-stripe 
I believe there are some cases, for example it's common practice AFAIK to annotate file name associated with a code block using such syntax:
````
```javascript markdoc.config.ts
// Some content
```
````

While we can surely achieve the same thing with `{% title="markdoc.config.ts" %}`, it could be annoying to type the attributes syxtax all the time.

Also the [CommonMark Spec](https://spec.commonmark.org/0.30/#info-string) referred to the entire part after ` ``` ` as info string, so I think having it avaliable to custom transform/render could be useful, and easy because it's already avaliable internally as `token.info` (I can PR if accepted). Users can decide whether to use Markdoc attribute syntax or to prefer their own conventional syntax of adding extra info.

**Reply by archiewood** [reply]
I'm also looking for this feature, and echo the above - while adding the annotations like this seems possible, it's somewhat ungainly.

Either of the following would work well for my use case
- exposing the full "info" string in the fence node
- exposing a "meta" attribute, which contains the info string minus the language part

Also happy to PR if this is felt to be sufficiently useful, else we can add this relatively easily to our fork.

**Reply by archiewood** [reply]
We've now published a fork with this feature, exposing a "meta" attribute as described above
- repo: https://github.com/evidence-dev/markdoc/tree/fence-meta-tags
- npm: https://www.npmjs.com/package/@archiewood/markdoc


**Usage:**

````
```sql some other string
select * from table
```
````

**Result:**
```
language: 'sql'
meta: 'some other string'
content: 'select * from table'
```
happy to PR in if more generally useful

**Comment by Lofty-Brambles** [comment]
That totally works for me! I didn't know that codeblocks could be used to add attributes like that!

**Reply by mfix-stripe** [reply]
@rpaul-stripe this seems like something we should document better in https://markdoc.dev/docs/nodes or something

**Reply by Pagebakers** [reply]
Took me a while to figure this out, would be great if this was added to the fence documentation.
